util/merkle-tree/src/merkle_tree.rs

```diff
@@ -0,0 +1,354 @@
+use super::merge;
+use numext_fixed_hash::H256;
+use serde_derive::{Deserialize, Serialize};
+use std::collections::VecDeque;
+
+pub fn merkle_root(leaves: &[H256]) -> H256 {
+    let leaves_len = leaves.len();
+    // in case of empty slice, just return zero
+    if leaves_len == 0 {
+        return H256::zero();
+    }
+
+    let mut queue = VecDeque::with_capacity((leaves_len + 1) >> 1);
+
+    if leaves_len & 1 == 1 {
```

Comment:
I prefer intuitive code style, e.g. we can use `rchunks_exact` here to replace the looping and modification of `i`
```
    let mut iter = leaves.rchunks_exact(2);
    while let Some([leaf1, leaf2]) = iter.next() {
        queue.push_back(merge(leaf1, leaf2))
    }
    if let [leaf] = iter.remainder() {
        queue.push_front(leaf.clone())
    }
```

Reply:
cool